kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <time.h>
+#include <errno.h>
+#include <mqueue.h>
+#include <string.h>
+
+struct mem_block {
+	struct k_mem_block block_name;
+	struct k_mem_block block_buffer;
+};
+
+typedef struct mqueue_object {
+	struct k_mem_block block_mqueue;
+	struct mem_block *block_buffer_name;
+	struct mqueue_object *next;
+	struct mqueue_object *prev;
+	struct k_msgq queue;
+	int ref_count;
+	char *name;
+} mqueue_object;
+
+typedef struct mqueue_desc {
+	struct k_mem_block block;
+	mqueue_object *mqueue;
+	u32_t  flags;
+} mqueue_desc;
+
+/* Define memory pool in heap area for message queues */
+#define MIN_BLK_SZ sizeof(struct mqueue_desc)
+#define MAX_BLK_SZ (CONFIG_MSG_COUNT_MAX * CONFIG_MSG_SIZE_MAX * sizeof(u8_t))
+
+#define SPACE_PER_MQ (MAX_BLK_SZ +	\
+		      CONFIG_MAX_DESC_PERMQ * sizeof(struct mqueue_desc) +\
+		      sizeof(struct mqueue_object) + sizeof(struct mem_block) +\
+		      CONFIG_MQUEUE_NAMELEN_MAX * sizeof(u8_t))
+
+#define MUL_FACTOR (SPACE_PER_MQ / MAX_BLK_SZ)
+#define CEL_MUL_FACTOR  ((MUL_FACTOR * MAX_BLK_SZ) < SPACE_PER_MQ ?	\
+			  MUL_FACTOR + 1 :	\
+			  MUL_FACTOR)
+K_MEM_POOL_DEFINE(posix_mqueue_pool, MIN_BLK_SZ, MAX_BLK_SZ,
+		  CEL_MUL_FACTOR * CONFIG_MQUEUE_QTY, 4);
```

Comment:
What's the design rationale for having this be a separate pool?  Seems like using k_malloc() for something like POSIX APIs that assume a system heap would be a better (and simpler) fit.

Reply:
agreed

Reply:
I changed it to memory pool based on feedback i received for pthread_create API. I was told to use memory pool.

If we agree, k_malloc is better approach then i will change this in my next patch set.